cpp/src/arrow/compute/kernels/scalar_list_benchmark.cc

```diff
@@ -0,0 +1,160 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <memory>
+#include <type_traits>
+#include <vector>
+
+#include "arrow/compute/api_scalar.h"
+#include "arrow/compute/exec.h"
+#include "arrow/testing/gtest_util.h"
+#include "arrow/testing/random.h"
+#include "arrow/util/benchmark_util.h"
+#include "benchmark/benchmark.h"
+
+namespace arrow::compute {
+
+constexpr auto kSeed = 0x94378165;
+
+const auto kSliceStart = 2;
+const auto kSliceStop = 10;
+
+template <typename InListType, typename ValueType, int32_t kFixedListSize = 0>
+static void BenchmarkListSlice(benchmark::State& state, const ListSliceOptions& opts) {
+  RegressionArgs args(state, /*size_is_bytes=*/false);
+  auto value_ty = TypeTraits<ValueType>::type_singleton();
+  std::shared_ptr<DataType> list_ty;
+  if constexpr (std::is_same_v<InListType, FixedSizeListType>) {
+    list_ty = std::make_shared<FixedSizeListType>(std::move(value_ty), kFixedListSize);
+  } else {
+    list_ty = std::make_shared<InListType>(std::move(value_ty));
+  }
+  auto rand = random::RandomArrayGenerator(kSeed);
+  auto array = rand.ArrayOf(list_ty, args.size, args.null_proportion);
+  auto ctx = default_exec_context();
+  std::vector<Datum> input_args = {std::move(array)};
+  for (auto _ : state) {
+    ABORT_NOT_OK(CallFunction("list_slice", input_args, &opts, ctx).status());
+  }
+}
+
+template <typename InListType = ListType>
+static void ListSliceInt64List(benchmark::State& state) {
+  ListSliceOptions opts;
+  opts.start = kSliceStart;
+  BenchmarkListSlice<InListType, Int64Type>(state, opts);
+}
+
+template <typename InListType = ListType>
+static void ListSliceStringList(benchmark::State& state) {
+  ListSliceOptions opts;
+  opts.start = kSliceStart;
+  BenchmarkListSlice<InListType, StringType>(state, opts);
+}
+
+template <typename InListType = ListType>
+static void ListSliceInt64ListWithStop(benchmark::State& state) {
+  ListSliceOptions opts;
+  opts.start = kSliceStart;
+  opts.stop = kSliceStop;
+  BenchmarkListSlice<InListType, Int64Type>(state, opts);
```

Comment:
Yes, because when `step=1`, the loops can operate on ranges instead of item-by-item.

And list-view implementations can benefit even more. I opened an issue for a more efficient `list_slice` kernel on list-views when `step=1`: #42066.

Reply:
I'm talking about stop here, not step.

Reply:
Sorry, I misread it (stop and step are too close). There are a bunch of branches on the presence of `stop` (`stop.has_value()`) so it's a relevant dimension for the kernel.

Having this on the benchmarks guided the refactoring and it's why the checks for has_value inside the loop are avoided in some situations.

Reply:
Thanks for the explanation, make sense.